confmap/confmap.go

```diff
@@ -537,7 +541,9 @@ type Marshaler interface {
 func zeroSliceHookFunc() mapstructure.DecodeHookFuncValue {
 	return func(from reflect.Value, to reflect.Value) (interface{}, error) {
 		if to.CanSet() && to.Kind() == reflect.Slice && from.Kind() == reflect.Slice {
-			to.Set(reflect.MakeSlice(to.Type(), from.Len(), from.Cap()))
+			if !from.IsNil() {
```

Comment:
is this check needed? I could see having this at the beginning:
```
		if !from.IsValid() {
			return from, nil
		}
```
But if the Value is valid then calling Len/Cap on nil slice should still be ok.

Reply:
@yurishkuro I agree with you that adding that block in the beginning is the way to go. It also fixes some other failures we were seeing. However, it causes one failure in `confmap_test.go`.
```
    --- FAIL: TestZeroSliceHookFunc/not_overridden_by_nil (0.00s)
        confmap_test.go:688: 
            	Error Trace:	D:/a/opentelemetry-collector/opentelemetry-collector/confmap/confmap_test.go:688
            	Error:      	Not equal: 
            	            	expected: &confmap.structWithSlices{Strings:[]string{"xxx", "yyyy"}}
            	            	actual  : &confmap.structWithSlices{Strings:[]string{}}
            	            	
            	            	Diff:
            	            	--- Expected
            	            	+++ Actual
            	            	@@ -1,5 +1,3 @@
            	            	 (*confmap.structWithSlices)({
            	            	- Strings: ([]string) (len=2) {
            	            	-  (string) (len=3) "xxx",
            	            	-  (string) (len=4) "yyyy"
            	            	+ Strings: ([]string) {
            	            	  }
            	Test:       	TestZeroSliceHookFunc/not_overridden_by_nil
```
Should this test expectation be changed now that we have enabled `DecodeNil`?

Reply:
No, there has to be a very high bar for changing test expectations, definitely not because we changed some implementation detail. Does the test make sense? Isn't it validating exactly what the hook promises to do?